app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,386 @@
+/**
+ * Profiling module that supports work with the [Web Performance API](https://developer.mozilla.org/en-US/docs/Web/API/Performance).
+ *
+ * Allows structured profiling of code with a hierarchical API.
+ *
+ * Example usage
+ * -------------
+ * ```
+ * // Manually start and end the measurement.
+ * const profilingHandle = profiling.task.start(doSomething')
+ * doSomething()
+ * profilingHandle.end()
+ *
+ * // Or use the `measure` method.
+ * profiling::task::measure("doSomething", doSomething);
+ *
+ * ```
+ *
+ * Note that this API and encoding formats for messages are synced with the rust equivalent in
+ * `lib/rust/profiling/src/lib.rs`.
+ */
+import * as assert from 'assert'
+
+// =========================
+// === Profiler Registry ===
+// =========================
+
+/// List of all instantiated profilers.
+const profilerRegistry: Profiler[] = []
+
+// ================
+// === Metadata ===
+// ================
+
+type SourceLocation = {
+    file: String
+    line: number
+}
+
+type Metadata = {
+    source: SourceLocation
+    profiling_level: string
+    label: string
+}
+
+/**
+ * Create a Metadata object that gets populated with the `SourceLocation` of the caller of the
+ * caller of this methods.
+ */
+function makeMetadata(profiling_level: string, label: string): Metadata {
+    const source = callerInfo(2)
+    return {
+        source,
+        profiling_level,
+        label,
+    }
+}
+
+/**
+ * Return the `SourceLocation` of this method call. By supplying the `caller_level` argument, the
+ * `SourceLocation` will be shifted up the call stack by that number of levels.
+ *
+ */
+function callerInfo(caller_level = 0): SourceLocation | null {
+    const e = new Error()
+    const regex = /\((.*):(\d+):(\d+)\)$/
+    const caller = e.stack.split('\n')[2 + caller_level]
+    if (caller === undefined) {
+        return null
+    }
+    const match = regex.exec(caller)
+    if (match === null) {
+        return null
+    }
+    assert(match.length >= 3, 'Regex was expected to produce three matches.')
```

Comment:
This error is not explanatory. Imagine that something breaks and user sees this error in a console. This would not be understandable at all.

Reply:
Also, this is a very fragile design. This line checks invariants for the upcoming lines. After refactoring things can change and break. Instead, use named groups in the pattern match!